quafu/qfasm/qfasm_utils.py

```diff
@@ -0,0 +1,133 @@
+import quafu.elements.element_gates as qeg
+
+class Node:
+    """
+    Node created in reduction.
+    Help for building QuantumCrcuit.
+    """
+    def __init__(self):
+        self.type = None
+        self.name = None
+        self.num = 0
+        self.start = 0
+        self.lineno = 0
+        self.filename = None
+    
+class UnaryExpr(Node):
+    def __init__(self, type, node:Node):
+        self.type = type
+        self.children = [node]
+
+class BinaryExpr(Node):
+    def __init__(self, type, left, right):
+        self.type = type
+        self.children = [left, right]
+
+class Id(Node):
+    def __init__(self, name, lineno, filename) :
+        self.name = name
+        self.lineno = lineno
+        self.filename = filename
+        self.num = 0
+
+class IndexedId(Node):
+    def __init__(self, node:Node, index):
+        self.num = index
+        self.name = node.name
+        self.lineno = node.lineno
+        self.filename = node.filename
+
+
+class GateInstruction:
+    def __init__(self, node, qargs, cargs = None, cbits = None):
+        self.name = node.name
+        self.lineno = node.lineno
+        self.filename = node.filename
+        self.qargs = qargs
+        self.cargs = cargs
+        self.cbits = cbits
+
+class IfInstruction:
+    def __init__(self, node, cbits, value:int, instruction):
+        self.name = node.name
+        self.lineno = node.lineno
+        self.filename = node.filename
+        self.cbits = cbits
+        self.value = value
+        self.instruction = instruction
+
+class SymtabNode(Node):
+    def __init__(self, type, node, is_global=True, is_qubit = False):
+        # qreg creg gate arg qarg
+        self.type = type
+        self.name = node.name
+        # for reg
+        self.num = 0
+        if isinstance(node, IndexedId):
+            self.num = node.num
+        self.start = 0
+        # for debug
+        self.lineno = node.lineno
+        self.filename = node.filename
+        # for check
+        self.is_global = is_global
+        # for gate
+        self.is_qubit = is_qubit
+        self.qargs = []
+        self.cargs = []
+        self.instructions = []
```

Comment:
Needs a blank line. Perhaps we can use [black](https://black.readthedocs.io/en/stable/) in the future.